components/brave_federated/adapters/flower_helper.cc

```diff
@@ -0,0 +1,205 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/adapters/flower_helper.h"
+
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "brave/components/brave_federated/task/model.h"
+#include "brave/components/brave_federated/task/typing.h"
+#include "brave/third_party/flower/src/brave/flwr/serde.h"
+#include "brave/third_party/flower/src/proto/flwr/proto/fleet.pb.h"
+#include "brave/third_party/flower/src/proto/flwr/proto/node.pb.h"
+
+namespace {
+// TODO(lminto): Create this wiki url
+// constexpr char kWikiUrl[] =
+//     "https://github.com/brave/brave-browser/wiki/Federated-Learning";
+
+}  // namespace
+
+namespace brave_federated {
+
+bool ValidatePullTaskInsResponse(const flower::PullTaskInsResponse response) {
+  if (response.task_ins_list_size() == 0) {
+    VLOG(3) << "No tasks received from FL service";
+    return false;
+  }
+
+  // TODO(lminto): Add multiple tasks validation
+  int max_tasks = 1;
+  for (int i = 0; i < max_tasks; i++) {
+    flower::TaskIns task_instruction = response.task_ins_list(i);
+
+    std::string id = task_instruction.task_id();
+    std::string group_id = task_instruction.group_id();
+    std::string workload_id = task_instruction.workload_id();
+    if (id.empty() || group_id.empty() || workload_id.empty()) {
+      VLOG(3) << "Invalid task id received from FL service";
+      return false;
+    }
+
+    if (!task_instruction.has_task()) {
+      VLOG(3) << "Task object is missing from task instruction";
+      return false;
+    }
+    flower::Task flower_task = task_instruction.task();
+
+    if (!flower_task.has_legacy_server_message()) {
+      VLOG(3) << "Server message is missing from task object";
+      return false;
+    }
+    flower::ServerMessage message = flower_task.legacy_server_message();
+
+    if (message.has_fit_ins()) {
+      if (GetVectorsFromParameters(message.fit_ins().parameters()).size() ==
+          0) {
+        VLOG(3) << "No parameters received from FL service";
+        return false;
+      }
+    } else if (message.has_evaluate_ins()) {
+      if (GetVectorsFromParameters(message.evaluate_ins().parameters())
+              .size() == 0) {
+        VLOG(3) << "No parameters received from FL service";
+        return false;
+      }
+    } else if (message.has_reconnect_ins()) {
+      VLOG(3) << "**: Legacy reconnect instruction received from FL service";
+      return false;
+    } else {
+      VLOG(3) << "**: Received unrecognized instruction from FL service";
+      return false;
+    }
+  }
+
+  return true;
+}
+
+TaskList ParseTaskListFromResponseBody(const std::string& response_body) {
+  flower::PullTaskInsResponse response;
+  if ((response.ParseFromString(response_body))) {
+    if (!ValidatePullTaskInsResponse(response)) {
```

Comment:
Hm, IMHO is an uncommon way to make a validation. Here we have 2 functions with that same logic (one for validation, the other for parsing). And now we have to synchronised all the changes.
Better to make do validation during the parsing (to have the one source of thrust).

Reply:
added also some tests for faulty task lists